driver/src/orderbook/streamed/state.rs

```diff
@@ -206,9 +209,37 @@ impl State {
             .get_address_by_id(order.buy_token)
             .ok_or_else(|| anyhow!("unknown buy token"))?;
         let buy_balance = self.balances.entry((user_id, buy_token)).or_default();
-        buy_balance_fn(buy_balance)?;
+        buy_balance_fn(buy_balance)
+    }
 
-        Ok(())
+    fn apply_solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<()> {
+        let fee_token = self
+            .tokens
+            .get_address_by_id(0)
+            .ok_or_else(|| anyhow!("solution without fee token"))?;
+        self.revert_last_solution(fee_token, block_batch_id);
+        self.last_solution = (block_batch_id, event.submitter, event.burnt_fees);
+        self.balances
+            .entry((event.submitter, fee_token))
+            .or_default()
+            .solution_submission(event.burnt_fees, block_batch_id)
+    }
+
+    fn revert_last_solution(&mut self, fee_token: TokenAddress, block_batch_id: BatchId) {
+        if self.last_solution.0 == block_batch_id {
+            // Neither unwrap can fail because we must have previously added the fee to the
+            // submitter in which case the balance must exist and be reversible.
+            self.balances
+                .get_mut(&(self.last_solution.1, fee_token))
+                .unwrap()
+                .revert_solution_submission(self.last_solution.2, block_batch_id)
+                .unwrap();
+            self.last_solution.2 = U256::zero();
```

Comment:
nit: I'd find a struct with named fields for last solution easier to read rather than having to look up .0,.1,.2.

Reply:
Can you just reuse the `SolutionSubmission` event struct?

Reply:
Agree it should be a struct. I made one instead of reusing so that we only store exactly what is needed.